src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp

```diff
@@ -932,9 +1002,385 @@ namespace
+                    }
+                }
+                else
+                {
+                    const float ratio =
+                        ec_lobe_ratio(
+                            rcp_eta,
+                            Favg_rcp_eta,
+                            Favg,
+                            Eavg_rcp_eta,
+                            Eavg,
+                            values->m_precomputed.m_reflection_weight,
+                            values->m_precomputed.m_refraction_weight);
+
+                    const float Eo = get_dir_albedo(
+                        mdf,
+                        rcp_eta,
+                        roughness,
+                        abs(cos_on));
+
+                    if (is_reflection)
+                    {
+                        // ratio(1/n) * (1 - E(1/n, uo)) * (1 - E(1/n, ui))
+                        // ------------------------------------------------
+                        //             Pi * (1 - Eavg(n))
+
+                        const float Ei = get_dir_albedo(
+                            mdf,
+                            rcp_eta,
+                            roughness,
+                            abs(cos_in));
+
+                        fms_multiplier =
+                            ratio * (1.0f - Eo) * (1.0f - Ei) / (Pi<float>() * (1.0f - Eavg));
+                    }
+                    else
+                    {
+                        // (1 - ratio(1/n)) * (1 - E(1/n, uo)) * (1 - E(n, ui))
+                        // ----------------------------------------------------
+                        //               Pi * (1 - Eavg(n))
+
+                        const float Ei = get_dir_albedo(
+                            mdf,
+                            eta,
+                            roughness,
+                            abs(cos_in));
+
+                        fms_multiplier =
+                            (1.0f - ratio) * (1.0f - Eo) * (1.0f - Ei) / (Pi<float>() * (1.0f - Eavg));
+                    }
+                }
+
+                fms_multiplier = lerp(0.0f, fms_multiplier, values->m_energy_compensation);
+                madd(value, fms, fms_multiplier);
+            }
+        }
     };
 
     typedef BSDFWrapper<GlassBSDFImpl, false> GlassBSDF;
+
+    const float MinEta = 1.01f;
+    const float MaxEta = 3.0f;
+
+    class GlassAlbedoTable
+      : public AlbedoTable3D
+    {
+      public:
+        GlassAlbedoTable(const float* table, const float min_eta, const float max_eta)
+          : AlbedoTable3D(table, min_eta, max_eta)
+        {
+        }
+
+        template <typename MDF>
+        GlassAlbedoTable(const MDF& mdf, const float min_eta, const float max_eta)
+          : AlbedoTable3D(min_eta, max_eta)
+        {
+            for (size_t z = 0; z < TableSize; ++z)
+            {
+                const float eta = lerp(m_min_eta, m_max_eta, static_cast<float>(z) / (TableSize - 1));
```

Comment:
Maybe `foundation::fit()` is a better... fit here?

Reply:
I find lerp easier to read than fit for this use case. And fix needs an extra static_cast<size_t>(0).